// zod-validation-error/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toValidationError = exports.fromZodError = exports.fromZodIssue = exports.errorMap = exports.isValidationErrorLike = exports.isValidationError = exports.ValidationError = void 0;
var ValidationError_1 = require("./ValidationError");
Object.defineProperty(exports, "ValidationError", { enumerable: true, get: function () { return ValidationError_1.ValidationError; } });
var isValidationError_1 = require("./isValidationError");
Object.defineProperty(exports, "isValidationError", { enumerable: true, get: function () { return isValidationError_1.isValidationError; } });
var isValidationErrorLike_1 = require("./isValidationErrorLike");
Object.defineProperty(exports, "isValidationErrorLike", { enumerable: true, get: function () { return isValidationErrorLike_1.isValidationErrorLike; } });
var errorMap_1 = require("./errorMap");
Object.defineProperty(exports, "errorMap", { enumerable: true, get: function () { return errorMap_1.errorMap; } });
var fromZodIssue_1 = require("./fromZodIssue");
Object.defineProperty(exports, "fromZodIssue", { enumerable: true, get: function () { return fromZodIssue_1.fromZodIssue; } });
var fromZodError_1 = require("./fromZodError");
Object.defineProperty(exports, "fromZodError", { enumerable: true, get: function () { return fromZodError_1.fromZodError; } });
var toValidationError_1 = require("./toValidationError");
Object.defineProperty(exports, "toValidationError", { enumerable: true, get: function () { return toValidationError_1.toValidationError; } });
