// @mui/toolpad/dist/exports/entrypoint.mjs
import {
  AppHostContext,
  ToolpadApp,
  componentsStore,
  pageComponentsStore
} from "./chunk-7FLN3QSR.mjs";
import "./chunk-CLVFJARY.mjs";
import "./chunk-SMH5LAMQ.mjs";

// src/entrypoint.tsx
import Button from "@mui/material/Button";
import * as React from "react";
import * as ReactDOM from "react-dom/client";
import createCache from "@emotion/cache";
import { CacheProvider } from "@emotion/react";
import { Box } from "@mui/material";
var IS_PREVIEW = process.env.NODE_ENV !== "production";
var IS_CUSTOM_SERVER = process.env.TOOLPAD_CUSTOM_SERVER === "true";
var cache = createCache({
  key: "css",
  prepend: true
});
cache.compat = true;
function setComponents(newComponents, pageComponents) {
  componentsStore.setState(newComponents);
  pageComponentsStore.setState(pageComponents);
}
var IS_RENDERED_IN_CANVAS = typeof window === "undefined" ? false : !!window.frameElement?.dataset?.toolpadCanvas;
var appHost = {
  isPreview: IS_PREVIEW,
  isCustomServer: IS_CUSTOM_SERVER,
  isCanvas: IS_RENDERED_IN_CANVAS
};
function Root({ ToolpadApp: ToolpadApp2, initialState, base }) {
  return /* @__PURE__ */ React.createElement(React.StrictMode, null, /* @__PURE__ */ React.createElement(CacheProvider, { value: cache }, /* @__PURE__ */ React.createElement(Button, { sx: { display: "none" } }), /* @__PURE__ */ React.createElement(AppHostContext.Provider, { value: appHost }, /* @__PURE__ */ React.createElement(ToolpadApp2, { basename: base, state: initialState })), /* @__PURE__ */ React.createElement(Box, { "data-testid": "page-ready-marker", sx: { display: "none" } })));
}
function init({ ToolpadApp: ToolpadApp2 = ToolpadApp, initialState, base }) {
  ReactDOM.createRoot(document.getElementById("root")).render(
    /* @__PURE__ */ React.createElement(Root, { base, ToolpadApp: ToolpadApp2, initialState })
  );
}
export {
  init,
  setComponents
};
